### DISCUSSION TITLE
Tag based navigation using extended attributes

### DISCUSSION THREAD

**Discussion by lytex** [alexluigit/dirvish#95] [Ideas]
Following a [discussion in r/emacs](https://www.reddit.com/r/emacs/comments/w16hmk/comment/igxa5iv/?context=3) I would like to talk about tag based navigation, using something like [extended attributes](https://en.wikipedia.org/wiki/Extended_file_attributes) (xattrs).

I think it's a good idea is to present them as "advanced bookmarks" and then exploring their possibilities.

## What are the benefits of using extended attributes over solutions like bookmark?
The difference with bookmarks is that bookmarks are like having one tag only, i.e. you can only search through all your bookmarks for the needed folder.

Having tags allows you to search across all your project files, or or your repos, or all your elisp folders which for example could include a folder with pdf learning resources and also repos with specific code you want to work on.

**Having multiple tags per folder**, and I think **this is the real deal**, allows you to write a "query" matching all folders that are repos in elisp, or all the learning resources about elisp (which will be the queries "repo" AND "elisp", "learning" AND "elisp" respectively)

You could as well have a tag named as your project (say "awesome_app") and find all folders of that project, or only repos of that project,  or only documentation of that project ("repo" AND "awesome_app", "docs" AND "awesome_app")

## File indexer
Another point is that if you want to track all the xattrs/tags no matter how they were created (e.g. through some external program, or through copying instead of only through dirvish) then a file indexer of some sort is necessary to discover them.

An implementation without a file indexer would be much simpler: only tags created within dirvish are stored in an internal database, but would lose tags created outside of it.

**Comment by alexluigit** [comment]
Thanks for your post, I'll have a look at it when I have a moment, before that happens I'll link the discussion section in the README so that more people will see this.  

> An implementation without a file indexer would be much simpler: only tags created within dirvish are stored in an internal database, but would lose tags created outside of it.

I would prefer implementing this without the external tools, meaning we tag files and read/write/query indices using elisp. If it works well, we can then consider adding a file indexer to retrieve the tags created outside of Dirvish.

**Reply by lytex** [reply]
By the way, I've just came across [this reddit post](https://www.reddit.com/r/emacs/comments/a7dkkj/does_dired_support_user_extended_attributes/):

There is a package to add tags inside of emacs called [BookmarkPlus](https://www.emacswiki.org/emacs/BookmarkPlus) that works with a extension of Dired called [DiredPlus](https://www.emacswiki.org/emacs/DiredPlus)

Essentially it's what we're talking about

**Reply by alexluigit** [reply]
That's great, so we can borrow something from them and don't have to write it from scratch.
